src/Build/BackEnd/BuildManager/BuildSubmission.cs

```diff
@@ -198,6 +198,13 @@ private void CheckForCompletion()
                 bool hasCompleted = (Interlocked.Exchange(ref _completionInvoked, 1) == 1);
                 if (!hasCompleted)
                 {
+                    // Did this submission have warnings elevated to errors? If so, mark it as
+                    // failed even though it succeeded (with warnings--but they're errors).
+                    if (((IBuildComponentHost)BuildManager).LoggingService.HasBuildSubmissionLoggedErrors(BuildResult.SubmissionId))
```

Comment:
Check for understanding:
The build is "completed" only after _completionEvent is set, so this is called before any build displays whether it has been completed. _completionInvoked is initially set to 0, so if the build has completed, checked via line 196, then it checks if we already knew the build was completed with the Interlocked.Exchange check, and if we hadn't already done completion-related things, we trigger them by setting _completionEvent. This check to see if we had logged errors without failing comes immediately before that, so before any BuildSubmission can be marked complete, it has to check whether it has logged any errors (or warnings converted to errors), hence fixing the bug. Is that right?

Reply:
To be a bit pedantic, you say "build" several times when it would be more correct to say "BuildSubmission". You can have multiple submissions in a single Begin/EndBuild pair, and that's done in practice in, e.g., the VS solution build--so that all of the different projects share project state across a VS build operation, but have their own Submissions (and so the project system can know when each project individually finishes and schedule the ones that depend on it).